crates/voicevox_core_c_api/src/compatible_engine.rs

```diff
@@ -359,6 +359,99 @@ pub unsafe extern "C" fn decode_forward(
     }
 }
 
+/// # Safety
+///
+/// - `f0`はRustの`&[f32; length as usize]`として解釈できなければならない。
+/// - `phoneme`はRustの`&[f32; phoneme_size * length as usize]`として解釈できなければならない。
+/// - `speaker_id`はRustの`&[i64; 1]`として解釈できなければならない。
+/// - `output`はRustの`&mut [f32; ((length + 2 * margin_width) * feature_dim) as usize]`として解釈できなければならない。
+#[unsafe(no_mangle)] // SAFETY: voicevox_core_c_apiを構成するライブラリの中に、これと同名のシンボルは存在しない
+pub unsafe extern "C" fn generate_full_intermediate(
+    length: i64,
+    phoneme_size: i64,
+    margin_width: i64,
+    feature_dim: i64,
+    f0: *mut f32,
+    phoneme: *mut f32,
+    speaker_id: *mut i64,
+    output: *mut f32,
+) -> bool {
+    init_logger_once();
+    assert_aligned(f0);
+    assert_aligned(phoneme);
+    assert_aligned(speaker_id);
+    assert_aligned(output);
+    let length = length as usize;
+    let phoneme_size = phoneme_size as usize;
+    let margin_width = margin_width as usize;
+    let feature_dim = feature_dim as usize;
+    let synthesizer = &*lock_synthesizer();
+    let result = ensure_initialized!(synthesizer).generate_full_intermediate(
+        length,
+        phoneme_size,
+        // SAFETY: The safety contract must be upheld by the caller.
+        unsafe { std::slice::from_raw_parts(f0, length) },
+        unsafe { std::slice::from_raw_parts(phoneme, phoneme_size * length) },
+        StyleId::new(unsafe { *speaker_id as u32 }),
+    );
+    match result {
+        Ok(output_arr) => {
+            // SAFETY: The safety contract must be upheld by the caller.
+            let output_slice = unsafe {
+                std::slice::from_raw_parts_mut(output, (length + 2 * margin_width) * feature_dim)
+            };
+            output_slice.clone_from_slice(&output_arr.into_raw_vec());
+            true
+        }
+        Err(err) => {
+            set_message(&format!("{err}"));
+            false
+        }
+    }
+}
+
+/// # Safety
+///
+/// - `audio_feature`はRustの`&[f32; (length * feature_dim) as usize]`として解釈できなければならない。
+/// - `speaker_id`はRustの`&[i64; 1]`として解釈できなければならない。
+/// - `output`はRustの`&mut [f32; length as usize * 256]`として解釈できなければならない。
+#[unsafe(no_mangle)] // SAFETY: voicevox_core_c_apiを構成するライブラリの中に、これと同名のシンボルは存在しない
+pub unsafe extern "C" fn render_audio_segment(
+    length: i64,
+    feature_dim: i64,
+    audio_feature: *mut f32,
+    speaker_id: *mut i64,
+    output: *mut f32,
+) -> bool {
+    init_logger_once();
+    assert_aligned(audio_feature);
+    assert_aligned(speaker_id);
+    let length = length as usize;
+    let feature_dim = feature_dim as usize;
+    let synthesizer = &*lock_synthesizer();
+    // SAFETY: The safety contract must be upheld by the caller.
+    let audio_feature_vec =
+        unsafe { std::slice::from_raw_parts(audio_feature, length * feature_dim) };
+    let result = ensure_initialized!(synthesizer).render_audio_segment(
+        ndarray::arr1(audio_feature_vec)
+            .into_shape([length, feature_dim])
+            .unwrap(),
```

Comment:
`generate_full_intermediate`と`render_audio_segment`の二つについては、折角入出力がndarrayになっているので、こうしてしまってよいと思います。

```suggestion
    let result = ensure_initialized!(synthesizer).render_audio_segment(
        // SAFETY: The safety contract must be upheld by the caller.
        unsafe {
            ndarray::ArrayView2::from_shape_ptr([length, feature_dim], audio_feature).to_owned()
        },
```

[`ArrayView::<_, _>::from_shape_ptr`](https://docs.rs/ndarray/0.16/ndarray/struct.ArrayBase.html#method.from_shape_ptr)の"Safety"の節に注意する必要がありますが、`render_audio_segment`の"Safety"はそのままでいいかなと。
(あえて変えるなら、`&[[f32; feature_dim]; length]`になるかなと)